File: programming_challenges/top3Words.js
// https://www.codewars.com/kata/most-frequently-used-words-in-a-text/python

/*
Write a function that, given a string of text (possibly with punctuation and
line-breaks), returns an array of the top-3 most occurring words, in 
descending order of the number of occurrences.

Assumptions:
    A word is a string of letters (A to Z) optionally containing one or more 
apostrophes (') in ASCII.
    Matches should be case-insensitive, and the words in the result should 
be lowercased.
    Ties may be broken arbitrarily.
    If a text contains fewer than three unique words, then either the top-2 
or top-1 words should be returned, or an empty array if a text contains no 
words.
*/

function top3Words (text) {

	// Clean the string

	// Create a Regular Expression that removes all instances\
	// of the following punctuation
    var re = new RegExp("[\\[\\]?*|;&_:<>%$§+|{}\\\\()@.,/]", 'g')
    text = text.replace(re, " ")
    // Separately remove triple primes
	text = text.replace("'''", " ");
	// --------------------------------------------------------

	// Create an object where each key-value pair represents a unique word\
	// from the string and its frequency

	// Create an array with all the words contained in the input\
	// string (start by splitting the string at spaces) (using\
	// a lowercase version of the string)
	init_words_list = text.toLowerCase().split(' ')
	// Object to contain a mapping of each word to its frequency
	word_counts = {}

	// Count the frequencies of each word

	// Loop through the items of init_words_list, and save the\
	// frequencies of each word (non-empty strings) in word_counts
	for (let i=0; i<init_words_list.length; i++) {
		var word = init_words_list[i];
		// If the item is not an empty string and the string is not\
		// just a prime, adjust the word frequency in word_counts
        if (word.length > 0 & word != "\'") {
			if (word in word_counts) {
				word_counts[word] += 1;
			} else {
				word_counts[word] = 1
			}
		}
	}

	// -----------------------------------------------------------

	// Get the three most frequent words in a single array, ordered by
	// the frequency of the words

	// Create an array that contains the keys (words) from word_counts
	var words_list = Object.keys(word_counts)

	// Create an array to hold the 3 most frequent words
	var max_frequencies = []

	// Loop through the list of words three times, to find the one with\
	// the highest frequency. At the end of each iteration, remove the\
	// selected word from the list so that in the next iteration we can\
	// find the most frequent word relative to the updated list

	// Repeat the following loop 3 times
	for (let i=0; i<3; i++) {
		// The current most frequent word
		var max_word = ''
		// The frequency of the current most frequent word
		var freq = 0
		// The index of the current most frequent word, relative\
		// to words_list
		var index = 0
		// The current length of words_list
		var words_left = words_list.length

		// Loop through the keys in word_counts and find the\
		// current most frequent word, given that we are only\
		// checking keys that are currently in words_list
		for (let j=0; j<words_left; j++) {
			// If we found a word that is more frequent than the\
			// one we had found so far, update the following\
			// variables
			if (word_counts[words_list[j]] > freq) {
				max_word = words_list[j]
				freq = word_counts[words_list[j]]
				index = j
			}
		}

        // If the max_word is not an empty string, add it\
        // to the array of most frequent words, and remove\
        // it from the array containing the keys of word_counts
        if (max_word) {
            max_frequencies.push(max_word)
            words_list.splice(index, 1)            
        }
	}
	// ----------------------------------------------------------------

	return max_frequencies

}

console.log("top3Words('  //wont won\'t won't ') =>", top3Words("  //wont won't won't "))
console.log('------------------------------')
console.log("top3Words(\"  '''  \") =>", top3Words("  '''  "))
console.log('------------------------------')
console.log("top3Words(\"  , e   .. \") =>", top3Words("  , e   .. "))
console.log('------------------------------')
console.log("top3Words(\"e e e e DDD ddd DdD: ddd ddd aa aA Aa, bb cc cC e e e\") =>", top3Words("e e e e DDD ddd DdD: ddd ddd aa aA Aa, bb cc cC e e e"))
console.log('------------------------------')
console.log("top3Words(\"  '  \") =>", top3Words("  '  "))
console.log('------------------------------')

File: programming_challenges/top_3_words.py
# https://www.codewars.com/kata/most-frequently-used-words-in-a-text/python

'''
Write a function that, given a string of text (possibly with punctuation and
line-breaks), returns an array of the top-3 most occurring words, in 
descending order of the number of occurrences.

Assumptions:
    A word is a string of letters (A to Z) optionally containing one or more 
apostrophes (') in ASCII.
    Matches should be case-insensitive, and the words in the result should 
be lowercased.
    Ties may be broken arbitrarily.
    If a text contains fewer than three unique words, then either the top-2 
or top-1 words should be returned, or an empty array if a text contains no 
words.
'''

def top_3_words(text):

    # Remove punctuation from the input string by replacing\
    # it with space (replace with space to facilitate the\
    # split later)
    for punctuation in "|\\!@#£$§%€&/{([)]=}?*-+,;.:-_«»<>":
        text = text.replace(punctuation, ' ')
    # Separately replace triple primes
    text = text.replace("'''", ' ')

    # Create a list with all the words found in the input\
    # string (split at spaces)
    words_list = text.split()
    
    # # Convert the words to lowercase and include only non-empty strings
    # and 1-character strings of a prime ("'")
    words_list = [str(word.lower()) for word in words_list if word not in "'"]

    # Create a dictionary to hold the frequencies for each unique word\
    # in the words_list list
    word_counts = {word:words_list.count(word) for word in set(words_list)}
    
    # List to contain the 3 most frequent words    
    max_frequencies = list()

    # Run this loop 3 times because we are looking for the 3\
    # most frequent words
    for i in range(3):

        # Variables to keep track of what's the most frequent\
        # word and its frequency
        max_word, freq = '', 0
        
        # Loop through the dictionary of words and its frequencies\
        # to find the most frequent
        for word in word_counts:
            # If this word has a higher frequency than the most\
            # frequent word found so far, it becomes the most\
            # frequent word
            if word_counts[word] > freq:
                max_word, freq = word, word_counts[word]
        
        # If the most frequent word we found is not an empty\
        # string, then add it to the list of the 3 most\
        # frequent words and delete it from the dictionary\
        # so that we can find the next most frequent word
        if max_word != '':
            max_frequencies.append(max_word)
            del word_counts[max_word]

        # If the list doesn't have any words then break\
        # the loop because it means we don't have any\
        # words to return
        if len(max_frequencies) == 0:
            break

    return max_frequencies


import unittest

class Test(unittest.TestCase):
    def test1(self):
        self.assertEqual(top_3_words("a a a  b  c c  d d d d  e e e e e"), ["e", "d", "a"])

    def test2(self):
        self.assertEqual(top_3_words("e e e e DDD ddd DdD: ddd ddd aa aA Aa, bb cc cC e e e"), ["e", "ddd", "aa"])

    def test3(self):
        self.assertEqual(top_3_words("  //wont won't won't "), ["won't", "wont"])

    def test4(self):
        self.assertEqual(top_3_words("  , e   .. "), ["e"])

    def test5(self):
        self.assertEqual(top_3_words("  ...  "), [])

    def test6(self):
        self.assertEqual(top_3_words("  '  "), [])

    def test7(self):
        self.assertEqual(top_3_words("  '''  "), [])

    def test8(self):
        self.assertEqual(top_3_words("""In a village of La Mancha, the name of which I have no desire to call to
mind, there lived not long since one of those gentlemen that keep a lance
in the lance-rack, an old buckler, a lean hack, and a greyhound for
coursing. An olla of rather more beef than mutton, a salad on most
nights, scraps on Saturdays, lentils on Fridays, and a pigeon or so extra
on Sundays, made away with three-quarters of his income."""), ["a", "of", "on"])


if __name__ == "__main__":
    unittest.main()